homeassistant/components/roborock/image.py

```diff
@@ -0,0 +1,128 @@
+
+    coordinators: dict[str, RoborockDataUpdateCoordinator] = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]
+    entities = list(
+        chain.from_iterable(
+            await asyncio.gather(
+                *(create_coordinator_maps(coord) for coord in coordinators.values())
+            )
+        )
+    )
+    async_add_entities(entities)
+
+
+class RoborockMap(RoborockCoordinatedEntity, ImageEntity):
+    """A class to let you visualize the map."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        unique_id: str,
+        coordinator: RoborockDataUpdateCoordinator,
+        map_flag: int,
+        starting_map: bytes,
+        map_name: str,
+    ) -> None:
+        """Initialize a Roborock map."""
+        RoborockCoordinatedEntity.__init__(self, unique_id, coordinator)
+        ImageEntity.__init__(self, coordinator.hass)
+        self._attr_name = map_name
+        self.parser = RoborockMapDataParser(
+            ColorsPalette(), Sizes(), IMAGE_DRAWABLES, ImageConfig(), []
+        )
+        self._attr_image_last_updated = dt_util.utcnow()
+        self.map_flag = map_flag
+        self.cached_map = self._create_image(starting_map)
+
+    def is_cache_expired(self) -> bool:
+        """Update this map if it is the current active map, it's been long enough, and the vacuum is cleaning."""
+        return (
+            self.map_flag == self.coordinator.current_map
+            and self.image_last_updated is not None
+            and (dt_util.utcnow() - self.image_last_updated).total_seconds()
+            > IMAGE_CACHE_INTERVAL
+            and self.coordinator.roborock_device_info.props.status is not None
+            and self.coordinator.roborock_device_info.props.status.in_cleaning
+        )
+
+    async def async_image(self) -> bytes | None:
+        """Update the image if it is not cached."""
+        if self.is_cache_expired():
+            map_data: bytes = await self.cloud_api.get_map_v1()
+            self.cached_map = self._create_image(map_data)
+            self._attr_image_last_updated = dt_util.utcnow()
+        return self.cached_map
+
+    def _create_image(self, map_bytes: bytes) -> bytes:
+        """Create an image using the map parser."""
+        parsed_map = self.parser.parse(map_bytes)
+        if parsed_map.image is None:
+            raise HomeAssistantError("Something went wrong creating the map.")
+        img_byte_arr = io.BytesIO()
+        parsed_map.image.data.save(img_byte_arr, format="PNG")
+        return img_byte_arr.getvalue()
+
+
+async def create_coordinator_maps(
+    coord: RoborockDataUpdateCoordinator,
+) -> list[RoborockMap]:
+    """Get the starting map information for all maps for this device. The following steps must be done synchronously.
+
+    Only one map can be loaded at a time per device.
+    """
+    entities = []
+    maps: MultiMapsList = await coord.cloud_api.get_multi_maps_list()
+    cur_map = coord.current_map
```

Comment:
Thanks, the comments are really helpful.

So technically, you could skip some calls here in cases where the current map is not changing and things are already loaded? Given there is a pretty huge sleep here, i'm wondering if there is a common case that has a fast path: For example, is it common to only have one map?

Reply:
Sorry to jump in but I'm also very interested to this great feature. 
To answer you last question I think many people have one map, but as soon as your house has 2 floors or there are stairs/steps the vacuum can't overcome, than you certainly have more than one map.

Reply:
Yeah I now check if we only have one map for this device and then if we do - I don't bother changing maps or sleeping

Reply:
Cool, this is great. You could also consider doing the current map last since i believe it needs to be restored at the end anyway. It might mean there are fewer places to have the multimap load call (just in the loop is ok)

Reply:
Got it, thank you.